File: project/core/module/rest-service/src/main/java/vdi/service/rest/util/BoundedInputStream.java
package vdi.service.rest.util;

import java.io.IOException;
import java.io.InputStream;
import java.util.function.Supplier;

public class BoundedInputStream extends InputStream {
  private final long maxBytes;

  private final InputStream inputStream;

  private final Supplier<RuntimeException> exceptionProvider;

  private long bytesRead;

  public BoundedInputStream(long maxBytes, InputStream inputStream, Supplier<RuntimeException> exceptionProvider) {
    this.maxBytes = maxBytes;
    this.inputStream = inputStream;
    this.exceptionProvider = exceptionProvider;
  }

  @Override
  public int read() throws IOException {
    if (++bytesRead > maxBytes)
      throw exceptionProvider.get();

    return inputStream.read();
  }

  @Override
  public int read(byte[] b) throws IOException {
    var red = inputStream.read(b);

    bytesRead += red;

    if (bytesRead > maxBytes)
      throw exceptionProvider.get();

    return red;
  }

  @Override
  public int read(byte[] b, int off, int len) throws IOException {
    var red = inputStream.read(b, off, len);

    bytesRead += red;
    if (bytesRead > maxBytes)
      throw exceptionProvider.get();

    return red;
  }

  @Override
  public void close() throws IOException {
    inputStream.close();
  }
}

File: project/core/module/rest-service/src/main/kotlin/vdi/service/rest/util/BoundedInputStream.kt
package vdi.service.rest.util

import java.io.InputStream

/**
 * Bounded Input Stream
 *
 * An input stream wrapper that limits the bytes read to the configured maximum.
 *
 * If the underlying [InputStream] contains more than the configured maximum
 * number of bytes, an exception will be thrown.  The thrown exception is
 * provided by a function given on instance construction.
 *
 * @author Elizabeth Paige Harper - https://github.com/foxcapades
 *
 * @constructor Constructs a new [BoundedInputStream] instance.
 *
 * @param maxBytes The maximum number of bytes that may be read from the
 * underlying input stream before an exception will be thrown.
 *
 * @param inputStream The underlying input stream that will be read from.
 *
 * @param exceptionProvider A provider function for the exception to throw in
 * the event that the underlying input stream contains more than the configured
 * max number of bytes.
 */
class BoundedInputStream(
  private val maxBytes: Long,
  private val inputStream: InputStream,
  private val exceptionProvider: () -> Throwable,
) : InputStream() {
  private var bytesRead = 0L

  override fun read(): Int {
    if (++bytesRead > maxBytes)
      throw exceptionProvider()

    return inputStream.read()
  }

  override fun read(b: ByteArray): Int {
    val red = inputStream.read(b)

    bytesRead += red
    if (bytesRead > maxBytes)
      throw exceptionProvider()

    return red
  }

  override fun read(b: ByteArray, off: Int, len: Int): Int {
    val red = inputStream.read(b, off, len)

    bytesRead += red
    if (bytesRead > maxBytes)
      throw exceptionProvider()

    return red
  }

  override fun close() {
    inputStream.close()
  }
}
